### ISSUE TITLE
The "Device ID does not match expected ID" error

### ISSUE THREAD

**Issue by JonyeeShen** [bendlabs/one_axis_ads#7]
I followed the guide provided by Sparkfun (https://learn.sparkfun.com/tutorials/nrf52840-development-with-arduino-and-circuitpython) and began by downloading the Arduino IDE (v2.1.0). In the "Additional Board Manager URL" field, I replaced the URL originally mentioned in the guide (https://github.com/adafruit/arduino-board-index/blob/gh-pages/package_adafruit_index.json) with https://adafruit.github.io/arduino-board-index/package_adafruit_index.json. The reason for this change was that the original URL led to "Arduino nRF52 Boards by Arduino", which did not align with the "Adafruit nRF52 by Adafruit" specified in the guide. After installing "Adafruit nRF52 by Adafruit" (v1.4.0), I successfully added "SparkFun Pro nRF52840 Mini" to the Board Manager by following the "Adding the SparkFun Pro nRF52840 Mini to the Board Menu" instructions in the Sparkfun nRF52840 guide.
	Next, I attempted to integrate the sensor and Sparkfun Pro nRF52840 Mini as per step 3 “Integrate Sensor and Sparkfun Pro nRF52840 Mini“ of the “Getting Started Guide”. However, I encountered an issue: the Serial Monitor displayed the error message "One Axis ADS initialization failed with reason: -5". Upon checking the error information, it appeared to indicate "Device ID does not match expected ID".
	I also tried following the SparkFun Displacement Sensor Arduino Library (GitHub - sparkfun/SparkFun_Displacement_Sensor_Arduino_Library: An Arduino library to interface to the angular displacement sensor (flex) from Bend Labs.), but was met with another error: "No sensor detected. Checking wiring. Freezing…".
	I would appreciate any guidance on potential solutions or similar problems that others may have encountered. Thank you.

**Comment by jendy12345** [comment]
Hello, I also encountered the same problem, but I still can't find the reason. My id address is 0x12, but it keeps saying that there is no match. Does anyone know what is causing this?

**Comment by donaldsaxby** [comment]
> ```
> I followed the guide provided by Sparkfun (https://learn.sparkfun.com/tutorials/nrf52840-development-with-arduino-and-circuitpython) and began by downloading the Arduino IDE (v2.1.0). In the "Additional Board Manager URL" field, I replaced the URL originally mentioned in the guide (https://github.com/adafruit/arduino-board-index/blob/gh-pages/package_adafruit_index.json) with https://adafruit.github.io/arduino-board-index/package_adafruit_index.json. The reason for this change was that the original URL led to "Arduino nRF52 Boards by Arduino", which did not align with the "Adafruit nRF52 by Adafruit" specified in the guide. After installing "Adafruit nRF52 by Adafruit" (v1.4.0), I successfully added "SparkFun Pro nRF52840 Mini" to the Board Manager by following the "Adding the SparkFun Pro nRF52840 Mini to the Board Menu" instructions in the Sparkfun nRF52840 guide.
> Next, I attempted to integrate the sensor and Sparkfun Pro nRF52840 Mini as per step 3 “Integrate Sensor and Sparkfun Pro nRF52840 Mini“ of the “Getting Started Guide”. However, I encountered an issue: the Serial Monitor displayed the error message "One Axis ADS initialization failed with reason: -5". Upon checking the error information, it appeared to indicate "Device ID does not match expected ID".
> I also tried following the SparkFun Displacement Sensor Arduino Library (GitHub - sparkfun/SparkFun_Displacement_Sensor_Arduino_Library: An Arduino library to interface to the angular displacement sensor (flex) from Bend Labs.), but was met with another error: "No sensor detected. Checking wiring. Freezing…".
> I would appreciate any guidance on potential solutions or similar problems that others may have encountered. Thank you.
> ```

Hi JonyeeShen, please contact me at dsaxby@bendlabs.com

**Comment by JonyeeShen** [comment]
@jendy12345 I solved this by changing the address in ads_driver to 0x15 manually, and it works.

**Comment by jendy12345** [comment]
@JonyeeShen  
Hello,
Do you use bend_address_update to update address when you update address? If yes, can you tell me what device ID you input for initialization? Thank you
![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/d3a59254-1761-4020-805a-454d3c67a4df)

Because I input new_address or device ID type wrong！
![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/01cc7429-dfa5-4e72-b784-bff5aa770146)
![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/b2543be3-8700-4d17-b496-e40c8db10f2e)

**Comment by JonyeeShen** [comment]
@jendy12345 Hello Jendy. I followed the steps below to make the sensor work. I am using a SparkFun Pro nRF52840 Mini with Arduino IDE.
1.	I manually changed ‘0x12’ in line 43 **ads_dfu.h** and line 17 **ads_hal_i2c.cpp** to ‘0x15’.
2.	I checked my device version following **bend_firmware_update_demo.ino** and ran it to have my firmware updated.
3.	I changed the ‘0x12’ in line 49 **bend_address_update.ino** to ‘0x15’ since the initial address in ads_driver has already been changed in former step 1. Then I ran this demo.
4.	In 2 and 3, I just pressed ENTER to serial monitor, and the sensor was automatically initialized.
5.	I uploaded the **bend_interrupt_demo.ino** with nothing changed and the sensor began working.
I hope the information above is helpful. If you have any further questions, you are welcome to contact me by 'junyi-shen[at]g.ecc.u-tokyo.ac.jp'

**Comment by jendy12345** [comment]
@JonyeeShen 
Hello,JonyeeShen.Is this your e-mail('junyi-shen[at]g.ecc.u-tokyo.ac.jp')?
I did it using arduino UNO board.I have tried to do what you said, but the problem is still the same.

The following is the part I modified：

1.![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/09eb60a0-8e92-4ce9-aef7-12a629edd550)

2.![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/2ae7e072-9830-43aa-b10b-d10cb23b119b)

3.
![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/bb1d7f51-1983-4368-b32a-95ecbbcb65a5)

May I ask where it has not been changed?

This is my error：
There is no response to pressing enter, and it is not correct to enter 0x15.

![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/d9de8956-9f32-4830-a748-54ab1e125003)

**Comment by JonyeeShen** [comment]
@jendy12345
  I have rechecked it by burning my board according to my former comment, and it seems fine to me. However, there are some possible reasons for your problem, 

	1. Since we are using different boards, you may have to change ‘#define ADS_RESET_PIN  (3)’ and ‘#define ADS_INTERRUPT_PIN  (4) ’ in demos according to your case because the default settings here are used for SparkFun board.
	2. Before I tried to update the address, I flashed the **bend_firmware_update_demo**; in this step, I checked my device version following lines 22-29 in this example. Remember to set your hardware version in either line 31 or 32 before you run it.

  Basically, I did it by changing ‘0x12’ to ‘0x15’ in ads_driver (3 places) -> checking my hardware version and having the firmware_update_demon run -> updating the address (it may not be necessary anymore since we have already updated it to 0x15 manually) -> flashing the bend_interrupt_demo
  I suggest you further check whether some more places need to change when using a different board if the steps above still fail to work.

**Comment by jendy12345** [comment]
@JonyeeShen thanks!!

**Comment by jendy12345** [comment]
@JonyeeShen Hello,excuse me,
I bought the sparkfun pro nrf52840 Mini board, but I can't find the corresponding board manager. How should I add the information of this board when running it on the arduino IDE?
I add this url
https://adafruit.github.io/arduino-board-index/package_adafruit_index.json has been added, but there is no information about the newly added SparkFun Pro nRF52840 Mini board. How should I add it? ? My arduino version is 1.8.19.
Thanks!

**Comment by JonyeeShen** [comment]
> @JonyeeShen Hello,excuse me, I bought the sparkfun pro nrf52840 Mini board, but I can't find the corresponding board manager. How should I add the information of this board when running it on the arduino IDE? I add this url https://adafruit.github.io/arduino-board-index/package_adafruit_index.json has been added, but there is no information about the newly added SparkFun Pro nRF52840 Mini board. How should I add it? ? My arduino version is 1.8.19. Thanks!

@jendy12345 Have you ever tried following “To add SparkFun nRF52840 board support to the nRF52 Arduino core, one file (boards.txt) needs to be modified, and one new folder (variants/sparkfun_nrf52840_mini) needs to be copied in …”, which is in the tutorial https://learn.sparkfun.com/tutorials/nrf52840-development-with-arduino-and-circuitpython?_gl=1*1py7u2l*_ga*MTE5MTcyMjQ0Ni4xNjg5NzY1NDQ0*_ga_T369JS7J9N*MTY4OTc3NTUxOS4yLjAuMTY4OTc3NTUxOS4wLjAuMA..&_ga=2.99241372.366362868.1689765445-1191722446.1689765444

**Comment by jendy12345** [comment]
@JonyeeShen ,Hello, I now have the corresponding name of the sparkfun_nRF52840_Mini board, but I have this problem now, have you ever encountered this problem?
![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/d964f773-e149-44cc-b642-9c99576c5ea1)
This error is reported when the program is compiled.
![image](https://github.com/bendlabs/one_axis_ads/assets/40566142/cc82eb60-4013-4359-bee1-32d324594099)

**Comment by jendy12345** [comment]
@JonyeeShen Hello,excuse me,because I haven't tested the nrf52840 board successfully, so can you please tell me the steps you did?
Thank you very much!

**Comment by JonyeeShen** [comment]
@jendy12345  Hello, I strictly followed this tutorial [nRF52840 Development with Arduino and CircuitPython - SparkFun Learn](https://learn.sparkfun.com/tutorials/nrf52840-development-with-arduino-and-circuitpython?_gl=1*e8cz6w*_ga*MTE5MTcyMjQ0Ni4xNjg5NzY1NDQ0*_ga_T369JS7J9N*MTY5MDQzMzY4My4zLjAuMTY5MDQzMzY4My42MC4wLjA.&_ga=2.140632306.1945468347.1690433685-1191722446.1689765444),  my Arduino IDE version is 2.1.0, and the board file version either 1.4.0 or 1.5.0 is okay based on my experience.

**Comment by jendy12345** [comment]
@JonyeeShen Hello,thank you!I downloaded the wrong version of board manager, now it can be used, thank you！
